src/operations/message_op.h

```diff
@@ -0,0 +1,162 @@
+/*******************************************************************************
+*   Ark Wallet
+*   (c) 2017 Ledger
+*   (c) ARK Ecosystem
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+********************************************************************************/
+
+#ifndef ARK_OPERATIONS_MESSAGE_H
+#define ARK_OPERATIONS_MESSAGE_H
+
+#include <stdint.h>
+
+#include <os.h>
+
+#include "constants.h"
+
+#include "operations/status.h"
+
+#include "utils/hex.h"
+#include "utils/print.h"
+
+#include "ux.h"
+#include "ux/display_context.h"
+
+////////////////////////////////////////////////////////////////////////////////
+
+// Sets the titles and variables to display a Message.
+//
+// Message must be: 0 < Msg <= 255
+//
+// -------
+// Display:
+//
+// Message > 0 && Message <= 64
+// - screen 1: Message Length
+// - screen 2: \"message\"
+//
+// ---
+//
+// Message > 64 && Message <= 128
+// - screen 1: Message Length
+// - screen 2: \"message...\"
+// - screen 3: \"...message\"
```

Comment:
I peeked at the code below but couldn't answer myself for sure, so:

It looks like this does not take into account the space occupied by the `...`. For example, assuming maximum length of the text that can be displayed is 64 chars and the message length is 125 - then at "screen 2" we would display 61 chars of the message + 3 chars for `...`. Then at "screen 3" we would display `...` + another 61 chars of the message (62..122). And we need "screen 4" to display the remaining chars from the message 123..125?

Do I miss something or this needs to be adjusted?

Reply:
I'll work on adding a better description of what's going on here.

The ellipses are always accounted for inside the loop,  
if only one screen is needed, the loop is skipped.

Every execution of the loop means there was necessarily a prior step.

every 2..4 step inside the loop:
- will always append `...` to the prior step
  - last var == screen 1: `msg + '...'`
  - last var == screen 2..3: `'...' + msg + '...'`
- necessarily begins with: `...`
- always sets its msg(`buffer[pos]`) at: `&var[step][3]`
- always null-terminates: `'...' + msg + '\0'`.

```c
    // Loop through the message steps,
    // prepending/appending ellipses where needed.
    // skipped if the Message only needs one screen.
    uint8_t step        = 0U;
    uint8_t pos         = 0U;
    for (uint8_t i = 0; i < steps - 2U; ++i) {
        step        = i + 2U;
        pos         =  HASH_64_LENGTH * (i + 1U);

        // - append ellipses to the end of the last var.
        os_memmove((char *)&displayCtx.var[i + 1U][HASH_64_LENGTH + (i ? 3U : 0U)],
                    "...", 4U);
...
        // - prepend the current var with ellipses.
        os_memmove((char *)displayCtx.var[step], "...", 3U);

        // - set the current var.
        os_memmove((char *)&displayCtx.var[step][3],
                    buffer + pos,
                    length - pos);

        // - null-terminate the prepended step var (... + var + \0)
        displayCtx.var[step][3U + length - pos] = '\0';
    }
```

---

this was the first implementation:

```c
    // - add ellipses to the end of the first var.
    // - set the second var.
    if (length > HASH_64_LENGTH) {
        os_memmove((char *)&displayCtx.var[1][HASH_64_LENGTH], "...", 4U);

        os_memmove((char *)displayCtx.title[2], "message pt 2:", 14U);
        os_memmove((char *)displayCtx.var[2], "...", 3U);

        os_memmove((char *)&displayCtx.var[2][3],
                   buffer + HASH_64_LENGTH,
                   length < HASH_64_LENGTH * 2U
                            ? length - HASH_64_LENGTH
                            : HASH_64_LENGTH * 2U);

        displayCtx.var[2][3U + length - HASH_64_LENGTH] = '\0';

        steps = 3U;
    }

    // if the buffer length > 128
    // - add ellipses to the end of the second var.
    // - set the third var.
    if (length > HASH_64_LENGTH * 2U) {
        os_memmove((char *)&displayCtx.var[2][HASH_64_LENGTH], "...", 4U);

        os_memmove((char *)displayCtx.title[3], "message pt 3:", 14U);
        os_memmove((char *)displayCtx.var[3], "...", 3U);

        os_memmove((char *)&displayCtx.var[3][3],
                   buffer + HASH_64_LENGTH * 2U,
                   length < HASH_64_LENGTH * 3U
                            ? length - HASH_64_LENGTH * 2U
                            : HASH_64_LENGTH * 3U);

        displayCtx.var[3][3U + length - HASH_64_LENGTH * 2U] = '\0';

        steps = 4U;
    }

    // if the buffer length > 192
    // - add ellipses to the end of the third var.
    // - set the fourth var.
    if (length > HASH_64_LENGTH * 3U) {
        os_memmove((char *)&displayCtx.var[3][HASH_64_LENGTH], "...", 4U);

        os_memmove((char *)displayCtx.title[4], "message pt 4:", 14U);
        os_memmove((char *)displayCtx.var[4], "...", 3U);

        os_memmove((char *)&displayCtx.var[4][3],
                   buffer + HASH_64_LENGTH * 3U,
                   length - HASH_64_LENGTH * 3U);

        displayCtx.var[4][3U + length - HASH_64_LENGTH * 3U] = '\0';

        steps = 5U;
    }
```